plugins/graalvm-nativeimage-emulation/src/main/java/org/qbicc/plugin/nativeimage/FeatureProcessor.java

```diff
@@ -0,0 +1,67 @@
+package org.qbicc.plugin.nativeimage;
+
+import com.oracle.svm.core.jdk.Resources;
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.impl.RuntimeClassInitializationSupport;
+import org.graalvm.nativeimage.impl.RuntimeReflectionSupport;
+import org.qbicc.context.CompilationContext;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+
+public class FeatureProcessor {
+    public static void processBuildFeature(CompilationContext ctxt, List<String> features, ClassLoader cl) {
+        if (features.isEmpty()) {
+            return;
+        }
+
+        // Fake up enough of the GraalVM Support classes to let us execute the
+        // beforeAnalysis method of the Feature and ex
+        Feature.BeforeAnalysisAccess qbiccProxy = new QbiccBeforeAnalysisAccess(ctxt);
+        QbiccImageSingletonsSupport qiss = new QbiccImageSingletonsSupport();
+        qiss.add(RuntimeReflectionSupport.class, new QbiccRuntimeReflectionSupport(ctxt));
+        qiss.add(RuntimeClassInitializationSupport.class, new QbiccRuntimeClassInitializationSupport(ctxt));
+        try {
+            Class<Resources> resourcesClass = (Class<Resources>)Class.forName("com.oracle.svm.core.jdk.Resources");
+            Constructor<Resources> rc = resourcesClass.getDeclaredConstructor();
+            rc.setAccessible(true);
+            Resources resourcesInstance = rc.newInstance();
+            qiss.add(Resources.class, resourcesInstance);
+        } catch (ClassNotFoundException | NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            ctxt.error("Unable to instantiate com.oracle.svm.core.jdk.Resources");
+            return;
+        }
+
+        // Set up the current Thread to load application classes into the hostVM via its CCL
+        ClassLoader savedCCL = Thread.currentThread().getContextClassLoader();
```

Comment:
I'm not super excited about this approach in the long term - I think it could collapse in certain cases (like when there are library or JDK conflicts or significant target platform disparity). In the short term though it's fine to get something off the ground. Hopefully in the future we can use VM invokables, perhaps combined with helper runtime classes, or a similar solution to retain the isolation between host and application.

Reply:
Yeah, I view this as a tactical approach to consume what Quarkus is currently generating.  If Quarkus makes progress on https://github.com/quarkusio/quarkus/issues/25975 and starts generating json files instead we could just read those json files directly instead of having to have a fake graalvm native image environment.